### ISSUE TITLE
Offline changes from multiple simulators/devices are not being sync as expected with automerge

### ISSUE THREAD

**Issue by aryunaabhay** [aws-amplify/amplify-flutter#634]
**Describe the bug**
with automerge activated offline changes are not being pushed when the server has a new version 

**To Reproduce**
1. Create a first version of the object on simulator A
2. Open simulator B and device C so they can obtain the initial version 
3. Turn off network for Device C, make changes on simulator B and wait till the changes are populated in dynamoDB
4. Then on deviceC execute the updates offline , close the app.
5. Turn on network and open the app so the offline changes are pushed, but these changes never arrive to dynamodb or the changes from the server are downloaded and the local changes are lost. 

**Expected behavior**
According to the automerge rules even though the Device C did the changes offline over a previous version, the changes on the individual properties should be able to be applied and generate a new version of the object in the server.

**Screenshots**
and when logging the datastore events the Device C when it reopens is supposedly syncing since it the modelSynced event is triggered, but at the end no data is pushed to the dynamodb and no error is presented.
<img width="382" alt="Screen Shot 2021-06-11 at 9 44 33 AM" src="https://user-images.githubusercontent.com/18670226/121704982-df4b6780-ca99-11eb-9654-c8db3f46f8a0.png">

**Platform**
Amplify Flutter current supports iOS and Android.  This issue is reproducible in (check all that apply):
[] Android
[x] iOS (I have tested it so far on iOS)

**Smartphone (please complete the following information):**
Simulators iphone 8 and iphone 12 using iOS 14.5
 - Device C: iPhone11
 - OS:  iOS14

**Comment by HuiSF** [comment]
@aryunaabhay Thanks for reporting this issue.

Do you recall, when you were configuring amplify api category, did you enable "conflict detection"? And what was the "default resolution strategy" you chose?

The options include
- Auto Merge 
- Optimistic Concurrency 
- Custom Lambda

**Comment by aryunaabhay** [comment]
Hi @HuiSF , yes I enabled conflict detection and I set the resolution strategy to "Auto merge".

**Comment by aryunaabhay** [comment]
I have created a script to run this test automatically on one computer with 3 simulators and is not working as expected either, I also have done these same steps manually but with no luck but today I did another manual test and it seemed to work, not sure why seems to work sometimes but no for most of the times. 

that takes me to the following doubts about how datastore offline strategy works : 

1. how much time takes for the datastore to send the changes done offline after reopening the app ? , 
2. does the datastore takes into account the date of the changes made offline ? 
3. also as I mentioned before in this sequence of steps the expected is just that automerge just creates a new version of the object in dynamodb merging the needed changes as they are stated here https://docs.aws.amazon.com/appsync/latest/devguide/conflict-detection-and-sync.html or am I missing some additional detail about how aws data store detects the conflicts and resolves them for the automerge mode ? 

cc: @HuiSF

**Comment by HuiSF** [comment]
Thanks much for the detailed information @aryunaabhay ! I will follow your findings, and start investigating the issue.

**Comment by aryunaabhay** [comment]
@HuiSF doing additional test , I think i've found the reason but it does not make much sense...
the object that i'm using for the test has a relationship of one to one with a smaller object and it seems that only the offline changes that also modify the connected object are the ones that are taken into account.

**Comment by HuiSF** [comment]
Sorry @aryunaabhay you are much ahead of me. 
It sounds like something went wrong when merge two versions of objects. I will dig deeper.
Please feel free to post your findings, I will follow up soon.

**Comment by HuiSF** [comment]
I did some digging, and testing. I couldn't reproduce the exact issue when doing test manually.

Here's the basic process of syncing remote and publishing local mutations

1. Start App (cold start)
2. Initial sync triggering 
3. Merging (if any updates were made, e.g. `__version` field update)
4. Once initial sync is completed, publishing locally queued mutations
5. Locally queued mutations are completed, newly updated data (returned from mutation requests) are written into local storage

I followed your steps did tests. Take `Post` as example. The issue seems reproducible in iOS but not in Android.

**In Android**

When device C is back online, it's synced remote posts first. This sync retrieved both versions of `Post`
1. The one created in device A (with `__version: 1`)
2. The one updated in device B (published when device B was back online, with `__version: 2`)

<details>
<summary>Complete sync json payload</summary>

```json
{
  "data": {
    "syncPosts": {
      "items": [
        {
          "_deleted": null,
          "_lastChangedAt": 1625166951105,
          "_version": 1,
          "author": {
            "id": "f77b4932-55ca-4ba3-af84-66044c46b1da"
          },
          "blog": {
            "id": "e0fe5957-d3c3-42e7-a072-3e6e8c00d55c"
          },
          "comments": {
            "items": [],
            "nextToken": null,
            "startedAt": null
          },
          "created": null,
          "id": "d4e5596a-b2e8-4680-8b4d-97ad910047dc",
          "rating": 10,
          "title": "Post 1"
        },
        {
          "_deleted": null,
          "_lastChangedAt": 1625167659374,
          "_version": 2,
          "author": {
            "id": "f77b4932-55ca-4ba3-af84-66044c46b1da"
          },
          "blog": {
            "id": "e0fe5957-d3c3-42e7-a072-3e6e8c00d55c"
          },
          "comments": {
            "items": [],
            "nextToken": null,
            "startedAt": null
          },
          "created": null,
          "id": "d4e5596a-b2e8-4680-8b4d-97ad910047dc",
          "rating": 10,
          "title": "Post 1 Modified from Pixel 4"
        }
      ],
      "nextToken": null,
      "startedAt": 1625168197352
    }
  }
}
```

</details>


After the sync, Android published the update mutation made when it was offline.
Note `__version: 2` in the mutation GraphQL document, 

```json
{
  "query": "mutation UpdatePost($input: UpdatePostInput!) {\n  updatePost(input: $input) {\n    _deleted\n    _lastChangedAt\n    _version\n    author {\n      id\n    }\n    blog {\n      id\n    }\n    comments {\n      items {\n        id\n      }\n      nextToken\n      startedAt\n    }\n    created\n    id\n    rating\n    title\n  }\n}\n",
  "variables": {
    "input": {
      "id": "d4e5596a-b2e8-4680-8b4d-97ad910047dc",
      "title": "Post 1 Modified from Pixel 3",
      "_version": 2
    }
  }
}
```

The mutation response is as following

```json
{
  "data": {
    "updatePost": {
      "_deleted": null,
      "_lastChangedAt": 1625168197917,
      "_version": 3,
      "author": {
        "id": "f77b4932-55ca-4ba3-af84-66044c46b1da"
      },
      "blog": {
        "id": "e0fe5957-d3c3-42e7-a072-3e6e8c00d55c"
      },
      "comments": {
        "items": [],
        "nextToken": null,
        "startedAt": null
      },
      "created": null,
      "id": "d4e5596a-b2e8-4680-8b4d-97ad910047dc",
      "rating": 10,
      "title": "Post 1 Modified from Pixel 3"
    }
  }
}
```

DynamoDB item is correctly updated. The `__version` field became `3`. 

**In iOS**

Everything looks very similar to Android. The only difference is at the update mutation GraphQL document. The `__version` field is still `1` even after the sync process.

```json
{
  "query": "mutation UpdatePost($input: UpdatePostInput!) {\n  updatePost(input: $input) {\n    id\n    created\n    rating\n    title\n    author {\n      id\n      email\n      name\n      department {\n        id\n        name\n        org {\n          id\n          name\n          __typename\n          _version\n          _deleted\n          _lastChangedAt\n        }\n        __typename\n        _version\n        _deleted\n        _lastChangedAt\n      }\n      __typename\n      _version\n      _deleted\n      _lastChangedAt\n    }\n    blog {\n      id\n      name\n      __typename\n      _version\n      _deleted\n      _lastChangedAt\n    }\n    __typename\n    _version\n    _deleted\n    _lastChangedAt\n  }\n}",
  "variables": {
    "input": {
      "id": "fade44af-26da-4494-a597-59b298d92af5",
      "rating": 10,
      "created": null,
      "title": "Post 1 Modified Pro",
      "authorId": "f77b4932-55ca-4ba3-af84-66044c46b1da",
      "blogID": "e0fe5957-d3c3-42e7-a072-3e6e8c00d55c",
      "_version": 1
    }
  }
}
```

This mutation update will cause conflict, which will be handled by "AutoMerge" following the "AutoMerge" scenario 1 listed in this [conflict detection documentation](https://docs.aws.amazon.com/appsync/latest/devguide/conflict-detection-and-sync.html) - the coming document has lower version, conflict fields would be dropped, and the `__version` field would be updated. Therefore, the update mutation response looks like following.

`title` field dropped the change made in device C when it was offline, `__version` field became 3.

```json
{
  "data": {
    "updatePost": {
      "id": "fade44af-26da-4494-a597-59b298d92af5",
      "created": null,
      "rating": 10,
      "title": "Post 1 Modified Pro Max",
      "author": {
        "id": "f77b4932-55ca-4ba3-af84-66044c46b1da",
        "email": null,
        "name": "Someone Testing",
        "department": {
          "id": "a97bc02e-c6dc-4fd4-86fd-6ce1d1953c7e",
          "name": "1st Writing",
          "org": {
            "id": "e416d068-3011-4321-bcc0-f5459f58a0eb",
            "name": "Writing Company",
            "__typename": "Organization",
            "_version": 1,
            "_deleted": null,
            "_lastChangedAt": 1624920912238
          },
          "__typename": "Department",
          "_version": 1,
          "_deleted": null,
          "_lastChangedAt": 1624920963365
        },
        "__typename": "Person",
        "_version": 1,
        "_deleted": null,
        "_lastChangedAt": 1624921012060
      },
      "blog": {
        "id": "e0fe5957-d3c3-42e7-a072-3e6e8c00d55c",
        "name": "Nature",
        "__typename": "Blog",
        "_version": 1,
        "_deleted": null,
        "_lastChangedAt": 1624920762067
      },
      "__typename": "Post",
      "_version": 3,
      "_deleted": null,
      "_lastChangedAt": 1625245421413
    }
  }
}
```

Hence, the local change made in device C when it was offline was actually published, and it made to DynamoDB, however, the expected change got dropped with AutoMerge. The behavior in Android looks correct. 

I will create an issue for `amplify-ios` to confirm and for further investigation.

**Comment by haverchuck** [comment]
@aryunaabhay  We are testing a [fix](https://github.com/aws-amplify/amplify-swift/pull/3402) for this issue within the amplify-swift repo.

**Comment by Equartey** [comment]
Hi @aryunaabhay, closing this as it has been fixed in [Amplify Flutter 1.7.0](https://pub.dev/packages/amplify_flutter/changelog#170).
